### PR TITLE
option to disable all scopes that `ActiveRecord.enum` generates

### LINKED ISSUE #34599: Add option to disable scopes in `ActiveRecord.enum`
The scopes generated by the `enum` feature is one of the things that causes conflicts with existing methods. Not everyone use those scopes so instead of renaming the enum value it would be better to handle those conflicts by not defining the scope.

```
enum {status: open}, _scopes: false
```

### PR DESCRIPTION
**Pull request by alfie-max**
### Summary

Fixes #34599 

By default `ActiveRecord.enum` generates several default scopes based on the values. This PR adds an option to disable those scopes which could possibly cause conflicts with existing methods for the model.

### CODE REVIEW

**Comment by alfie-max** [comment]
@rafaelfranca just gave this a try.. kindly check this out and maybe i can work on making some changes if necessary

**Comment by SamSaffron** [comment]
This looks correct to me but we should have a test I think, that way we can confirm it will not regress, params with the name `_enum` etc. end up being prime candidates long term for removal cause they do not feel quite right.

**Comment by alfie-max** [comment]
@SamSaffron added a test.. Not sure if that would suffice. Actually caught a case which i missed which would be the default case while adding the test.

File: activerecord/lib/active_record/enum.rb [hunk 1]
```diff
@@ -194,9 +195,11 @@ def enum(definitions)
             klass.send(:detect_enum_conflict!, name, "#{value_method_name}!")
             define_method("#{value_method_name}!") { update!(attr => value) }
 
-            # scope :active, -> { where(status: 0) }
-            klass.send(:detect_enum_conflict!, name, value_method_name, true)
-            klass.scope value_method_name, -> { where(attr => value) }
+            if enum_scopes.nil? || enum_scopes == true
```
[the remark this answers is not in the document]
Comment (gmcgibbon) [reply activerecord/lib/active_record/enum.rb]: I think you can just do `enum_scopes.nil? || enum_scopes`

File: activerecord/lib/active_record/enum.rb [hunk 2]
```diff
@@ -194,9 +195,11 @@ def enum(definitions)
...
+              # scope :active, -> { where(status: 0) }
```
Comment (gmcgibbon) [inline activerecord/lib/active_record/enum.rb]: This probably doesn't need to be here. 😄

File: activerecord/test/cases/enum_test.rb [hunk 3]
```diff
@@ -551,4 +551,13 @@ def self.name; "Book"; end
   test "data type of Enum type" do
     assert_equal :integer, Book.type_for_attribute("status").type
   end
+
+  test "default scopes disabled" do
```
Comment (gmcgibbon) [inline activerecord/test/cases/enum_test.rb]: `"scopes can be disabled"`?

**Review by gmcgibbon — APPROVED** [review]
👍 LGTM. Please squash your commits

**Comment by alfie-max** [comment]
Sure.. just a min

**Comment by alfie-max** [comment]
should i add this to the CHANGELOG?

File: activerecord/lib/active_record/enum.rb [hunk 4]
```diff
@@ -195,8 +196,10 @@ def enum(definitions)
             define_method("#{value_method_name}!") { update!(attr => value) }
 
             # scope :active, -> { where(status: 0) }
-            klass.send(:detect_enum_conflict!, name, value_method_name, true)
-            klass.scope value_method_name, -> { where(attr => value) }
+            if enum_scopes.nil? || enum_scopes
```
Comment (rafaelfranca) [inline activerecord/lib/active_record/enum.rb]: Better to change this to `if enum_scopes != false`

Comment (alfie-max) [reply activerecord/lib/active_record/enum.rb]: ok.. will do

**Comment by rafaelfranca** [comment]
👍 for a changelog entry

**Review by gmcgibbon — APPROVED** [review]

File: activerecord/test/cases/enum_test.rb [hunk 5]
```diff
@@ -551,4 +551,13 @@ def self.name; "Book"; end
...
     assert_equal :integer, Book.type_for_attribute("status").type
   end
+
+  test "scopes can be disabled" do
+    klass = Class.new(ActiveRecord::Base) do
+      self.table_name = "books"
+      enum status: [:proposed, :written], _scopes: false
+    end
+
+    assert_raises(NoMethodError) { klass.proposed? }
```
Comment (SamSaffron) [inline activerecord/test/cases/enum_test.rb]: Shouldn't this be `klass.proposed` ? which is the scope?

Comment (alfie-max) [reply activerecord/test/cases/enum_test.rb]: fixed it.. 😄

**Comment by rafaelfranca** [comment]
Thank you so much!

**Comment by SamSaffron** [comment]
congrats @alfie-max on getting your first Rails PR done! Thanks

**Comment by alfie-max** [comment]
thanks @SamSaffron  😄

**Pull request merged**